crispr_read_counts/single_guide_count.py

```diff
@@ -0,0 +1,249 @@
+    except Exception as e:
+      sys.exit(error_msg('Unexpected exception when trying to open input CRAM file: %s' % str(e)))
+
+    for rg in samfile.header.to_dict().get('RG'):  # does not matter which RG line's SM tag is used
+      self.sample_name = rg.get('SM')
+    if not self.sample_name:
+      sys.exit(error_msg('Could not find "SM" tag in the input file header'))
+    return samfile
+
+  def get_lib_seq_dict_and_seq_length(self, reverse_complementing):
+    lib_seqs = {}
+    for seq in self.lib.keys():
+      key = seq
+      if reverse_complementing:
+        # reverse complementing guide RNA sequences instead of each read
+        key = rev_compl(key)
+      lib_seqs[key] = seq
+
+    # assume library sequences are in same length
+    for seq in lib_seqs.keys():
+      lib_seq_size = len(seq)
+      break
+
+    return lib_seqs, lib_seq_size
+
+  def _get_sgrna_library_counts_with_stats(self, trim: int, reverse_complementing: bool):
+
+    total_reads, vendor_failed_reads, mapped_to_guide_reads = 0, 0, 0
+
+    samfile = self.open_cram_and_get_sample_name()
+    lib_seqs, lib_seq_size = self.get_lib_seq_dict_and_seq_length(reverse_complementing)
+
+    for read in samfile.fetch(until_eof=True):
+      # if the alignment is secondary or supplymentary, skip it!
+      if read.flag & 2304:
+        continue
+
+      # if the alignment is vendor failed, skip it but count it!
+      if read.flag & 512:
+        total_reads += 1
+        vendor_failed_reads += 1
+        continue
+
+      total_reads += 1
+      if reverse_complementing:
+        # NOTE: it could be as simple as this: cram_seq = read.get_forward_sequence()[-trim-lib_seq_size:-trim]
+        # but when trim is 0, end index will be "-0", and this will upset python and return an empty string.
+        # so, trim is always added by 1, and the substring became a bit complicated like below.
+        cram_seq = read.get_forward_sequence()
+        cram_seq = cram_seq[-(trim + 1) - (lib_seq_size - 1):-(trim + 1)] + cram_seq[-(trim + 1)]
+      else:
+        cram_seq = read.get_forward_sequence()[trim:trim + lib_seq_size]
+
+      matching_lib_seq = lib_seqs.get(cram_seq)
+      if matching_lib_seq:
+        mapped_to_guide_reads += 1
+        for grna_id in self.lib[matching_lib_seq]:
+          self.sample_count[grna_id] = self.sample_count.get(grna_id, 0) + 1
+
+    self.stats['total_reads'] = total_reads
+    self.stats['vendor_failed_reads'] = vendor_failed_reads
+    self.stats['mapped_to_guide_reads'] = mapped_to_guide_reads
+
+  def _get_sgrna_library_counts_without_stats(self, trim: int, reverse_complementing: bool):
+
+    samfile = self.open_cram_and_get_sample_name()
+    lib_seqs, lib_seq_size = self.get_lib_seq_dict_and_seq_length(reverse_complementing)
+
+    for read in samfile.fetch(until_eof=True):
+      # if the alignment is secondary, supplymentary or  vendor failed, skip it!
+      if read.flag & 2816:
+        continue
+
+      if reverse_complementing:
```

Comment:
Reverse slicing is a corner case that has annoyed me before, so I investigated a bit...

I believe the following could be easier to digest and possibly faster (to be thoroughly checked).

Before the loop:

```python
sl = (
    slice(trim, trim + lib_seq_size) if not reverse_complementing else
    slice(-trim - lib_seq_size, -trim) if trim > 0 else
    slice(-lib_seq_size, None)
)
```

In the loop:

```python
cram_seq = read.get_forward_sequence()[sl]
```

Reply:
Wow, this is wonderful!